draft-ietf-jose-hpke-encrypt.md

```diff
@@ -131,70 +131,172 @@ This specification uses the following abbreviations and terms:
+  *  Key Encryption, where HPKE is used to encrypt a content encryption key (CEK) for encrypting the plaintext.
 
-When "alg" and "enc" are both present in a protected header and when "iv" and "tag" are empty, the mode is HPKE JWE Integrated Encryption.
+When "alg" is a JOSE-HPKE algorithm:
 
-When "enc" is present in a protected header and "alg" is absent, the mode is HPKE JWE Key Encryption when a valid HPKE "alg" value is present in the unprotected headers.
+  * If "enc" is "dir", HPKE JWE Integrated Encryption is used.
+  * If "enc" is an AEAD algorithm, The recipient Key Managment mode is Key Encryption.
+
+The HPKE KEM, KDF and AEAD used depend on the JOSE-HPKE algorithm used.
 
 HPKE supports several modes, which are described in Table 1 of {{RFC9180}}.
 
-In JWE, the use of specific HPKE modes such as "mode_base" or "mode_auth_psk" is determined by the presence of the header parameters "psk_id" and "auth_kid".
+The HPKE mode used is determined by header parameters "psk_id" and "auth_kid" in the JOSE Header:
+
+  * If neither is present, base mode is used.
+  * If only "psk_id" is present, PSK mode is used.
+  * If only "auth_kid" is present,  Auth mode is used.
+  * If both are present, AuthPSK mode is used.
+
+If authentication is not supported by the HPKE KEM used, the "auth_kid" header parameter MUST be absent.
+
+
+## HPKE JWE Integrated Encryption
+
+When encrypting, the inputs to HPKE Seal operation are set as follows:
+
+- kem_id: Depends on the JOSE-HPKE algorithm used.
+- pkR: The recipient public key, converted into HPKE public key.
+- kdf_id: Depends on the JOSE-HPKE algorithm used.
+- aead_id: Depends on the JOSE-HPKE algorithm used.
+- info: By default, an empty string. Application MAY specify some other value.
+- aad: Additional Authenticated Data encryption parameter defined in {{RFC7516}} section 5.1 step 14.
+- pt: The message plaintext, compressed using the algorithm in "zip" header parameter if present.
+
+Additionally, if the "psk_id" header parameter is present:
+
+- psk: The pre-shared key.
+- psk_id: The value of the "psk_id" header parameter.
+
+Additionally, if the "auth_kid" header parameter is present:
+
+- skS: The sender private key, converted into HPKE private key.
+
+The resulting JWE is filled as follows:
+
+- JWE Protected Header:
+  * SHALL contain "enc" with value "dir"
+  * SHALL contain "alg" that is the used JOSE-HPKE algorithm.
+  * MUST contain the "apu", "apv" and "zip" header parameters, if present.
+- JWE Initialization Vector SHALL be empty.
+- JWE Ciphertext SHALL be the raw ct output from HPKE Seal operation.
+- JWE Authentication Tag SHALL be empty.
+- There MUST be exactly one recipient, with:
+  * JWE Per-Recipient Unprotected Header MUST be empty.
+  * JWE Encrypted Key SHALL contain the raw enc output from HPKE Seal operation.
+
+The "ek" header parameter MUST NOT be present.
+
+When decrypting, the inputs to HPKE Open operation are set as follows:
```

Comment:
That is to leave the HPKE setup info free for application to use for actual context binding.  Apu and apv in JOSE are pretty much completely broken.

Reply:
If apu and apv is broken, why use it for key encryption mode only ?

Reply:
I included because it was there before, I can only speculate why it was there.

Reply:
I prefer including them in both modes to maintain consistency with Section 4.6.2 of RFC 7518, or alternatively, not including them with an explanation that it is broken.

Reply:
Section 4.6.2 is only applicable to Key Agreement, which HPKE is not (the internal structure of HPKE is irrelevant here).

Consistency would be to other Key Encryption algorithms, which do not support apu nor apv. The "enc" ensures that cross-mode attack is not possible, and that bad JWE implementations are not exposed to oracle attack.

AFAICT, apu and apv are just useless in JWE. The arguments for those only hold for ECDH-SS, and the section 4.6.2 context structure would not work for ECDH-SS (no place to put the nonce required for security).

Reply:
I don't see any specification stating that apu and apv are useless in JWE. Until there is consensus in the WG, we will need to include apu and apv in both modes. As an alternative, we could consider explicitly documenting the decision to exclude apu and apv and explain that they are not necessary for HPKE in JWE. This would provide clarity to implementers and trigger further discussion within the WG

Reply:
No current Key Encryption algorithm supports apu nor apv. That would be one reason to exclude those.

Reply:
As per Section 4.6.2 of RFC 7518, both apu and apv are used to derive the key which is used to encrypt the CEK.

Reply:
I will raise an issue on GitHub and start a thread on the JOSE WG mailing list after the holidays.